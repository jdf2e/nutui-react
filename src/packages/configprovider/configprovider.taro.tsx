import React, { FunctionComponent, createContext, useContext } from 'react'
import classNames from 'classnames'
import kebabCase from 'lodash.kebabcase'
import isequal from 'lodash.isequal'
import useMemo from '@/utils/use-memo'
import { BasicComponent } from '@/utils/typings'
import { BaseLang } from '@/locales/base'
import zhCN from '@/locales/zh-CN'
import type { NutCSSVariables } from './types'

export interface ConfigProviderProps extends BasicComponent {
  locale: BaseLang
  direction?: ConfigProviderDirection
  theme?: Record<string | NutCSSVariables, string>
}

export type ConfigProviderDirection = 'ltr' | 'rtl' | undefined

const classPrefix = 'nut-configprovider'

export const defaultConfigRef: {
  current: ConfigProviderProps
} = {
  current: {
    locale: zhCN,
    direction: 'ltr',
  },
}

export const setDefaultConfig = (config: ConfigProviderProps) => {
  defaultConfigRef.current = config
}

export const getDefaultConfig = () => {
  return defaultConfigRef.current
}

// 创建一个 Context 对象
const ConfigContext = createContext<ConfigProviderProps | null>(null)

export const useConfig = () => {
  return useContext(ConfigContext) ?? getDefaultConfig()
}

function convertThemeVarsToCSSVars(themeVars: Record<string, string | number>) {
  const cssVars: Record<string, string | number> = {}
  Object.keys(themeVars).forEach((key) => {
    cssVars[`--${kebabCase(key)}`] = themeVars[key]
  })
  return cssVars
}

export const useRtl = () => {
  const { direction } = useConfig()
  return direction === 'rtl'
}

export const ConfigProvider: FunctionComponent<
  Partial<ConfigProviderProps & BasicComponent>
> = (props) => {
  const { style, className, children, direction, ...config } = props

  const mergedConfig = useMemo(
    () => {
      return {
        ...getDefaultConfig(),
        ...config,
        direction,
      }
    },
    [config, direction],
    (prev, next) =>
      prev.some((prevTheme, index) => {
        const nextTheme = next[index]

        return !isequal(prevTheme, nextTheme)
      })
  ) as ConfigProviderProps

  const cssVarStyle = React.useMemo(() => {
    return convertThemeVarsToCSSVars(mergedConfig.theme || {})
  }, [mergedConfig.theme])

  return (
    <ConfigContext.Provider value={mergedConfig}>
      <div
        className={classNames(classPrefix, className, `nut-${direction}`)}
        style={{
          ...cssVarStyle,
          ...style,
          direction,
        }}
      >
        {children}
      </div>
    </ConfigContext.Provider>
  )
}

ConfigProvider.displayName = 'NutConfigProvider'
