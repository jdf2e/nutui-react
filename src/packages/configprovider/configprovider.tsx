import React, { FunctionComponent, createContext, useContext } from 'react'
import classNames from 'classnames'
import kebabCase from 'lodash.kebabcase'
import isequal from 'lodash.isequal'
import useMemo from '@/utils/use-memo'
import { BasicComponent } from '@/utils/typings'
import { BaseLang } from '@/locales/base'
import zhCN from '@/locales/zh-CN'
import type { NutCSSVariables } from './types'

export interface ConfigProviderProps extends BasicComponent {
  locale: BaseLang
  theme?: Record<string | NutCSSVariables, string>
}

const classPrefix = 'nut-configprovider'

export const defaultConfigRef: {
  current: ConfigProviderProps
} = {
  current: {
    locale: zhCN,
  },
}

export const setDefaultConfig = (config: ConfigProviderProps) => {
  defaultConfigRef.current = config
}

export const getDefaultConfig = () => {
  return defaultConfigRef.current
}

// 创建一个 Context 对象
const ConfigContext = createContext<ConfigProviderProps | null>(null)

export const useConfig = () => {
  return useContext(ConfigContext) ?? getDefaultConfig()
}

function convertThemeVarsToCSSVars(themeVars: Record<string, string | number>) {
  const cssVars: Record<string, string | number> = {}
  Object.keys(themeVars).forEach((key) => {
    cssVars[`--${kebabCase(key)}`] = themeVars[key]
  })
  return cssVars
}

export const ConfigProvider: FunctionComponent<
  Partial<ConfigProviderProps & BasicComponent>
> = (props) => {
  const { style, className, children, ...config } = props

  const mergedConfig = useMemo(
    () => {
      return {
        ...getDefaultConfig(),
        ...config,
      }
    },
    [config],
    (prev, next) =>
      prev.some((prevTheme, index) => {
        const nextTheme = next[index]

        return !isequal(prevTheme, nextTheme)
      })
  ) as ConfigProviderProps

  const cssVarStyle = React.useMemo(() => {
    return convertThemeVarsToCSSVars(mergedConfig.theme || {})
  }, [mergedConfig.theme])

  return (
    <ConfigContext.Provider value={mergedConfig}>
      <div
        className={classNames(classPrefix, className)}
        style={{
          ...cssVarStyle,
          ...style,
        }}
      >
        {children}
      </div>
    </ConfigContext.Provider>
  )
}

ConfigProvider.displayName = 'NutConfigProvider'
