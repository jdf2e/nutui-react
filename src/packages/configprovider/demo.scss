.nut-configprovider-cell-title {
  font-size: 14px;
}
.nut-configprovider-cell-desc {
  font-size: 12px;
}
