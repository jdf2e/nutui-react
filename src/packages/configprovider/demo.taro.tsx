import React from 'react'
import Taro from '@tarojs/taro'
import Header from '@/sites/components/header'
import Demo1 from './demos/taro/demo1'
import Demo2 from './demos/taro/demo2'
import Demo3 from './demos/taro/demo3'
import Demo4 from './demos/taro/demo4'
import Demo5 from './demos/taro/demo5'
import { useTranslate } from '@/sites/assets/locale/taro'

const ConfigProviderDemo = () => {
  const [translated] = useTranslate({
    'zh-CN': {
      title1: 'Textarea 默认',
      title2: 'Textarea 英文',
      defaultTheme: '默认主题',
      customTheme: '定制主题',
    },
    'zh-TW': {
      title1: '默認用法',
      title2: 'Textarea 英文',
      defaultTheme: '默認主題',
      customTheme: '定制主題',
    },
    'en-US': {
      title1: 'Textarea default',
      title2: 'Textarea en-US',
      customTheme: 'Custom Theme',
      defaultTheme: 'Default Theme',
    },
  })
  return (
    <>
      <Header />
      <div className={`demo ${Taro.getEnv() === 'WEB' ? 'web' : ''}`}>
        <h2>{translated.defaultTheme}</h2>
        <Demo1 />
        <h2>{translated.customTheme}</h2>
        <Demo2 />
        <h2>{translated.title1}</h2>
        <Demo3 />
        <h2>{translated.title2}</h2>
        <Demo4 />
        <h2>RTL</h2>
        <Demo5 />
      </div>
    </>
  )
}

export default ConfigProviderDemo
