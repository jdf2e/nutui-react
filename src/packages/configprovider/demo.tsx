import React from 'react'
import Demo1 from './demos/h5/demo1'
import Demo2 from './demos/h5/demo2'
import Demo3 from './demos/h5/demo3'
import Demo4 from './demos/h5/demo4'
import Demo5 from './demos/h5/demo5'
import { useTranslate } from '@/sites/assets/locale'

const ConfigProviderDemo = () => {
  const [translated] = useTranslate({
    'zh-CN': {
      title1: 'Textarea 默认',
      title2: 'Textarea 英文',
      defaultTheme: '默认主题',
      customTheme: '定制主题',
    },
    'zh-TW': {
      title1: '默認用法',
      title2: 'Textarea 英文',
      defaultTheme: '默認主題',
      customTheme: '定制主題',
    },
    'en-US': {
      title1: 'Textarea default',
      title2: 'Textarea en-US',
      customTheme: 'Custom Theme',
      defaultTheme: 'Default Theme',
    },
  })
  return (
    <>
      <div className="demo">
        <h2>{translated.defaultTheme}</h2>
        <Demo1 />
        <h2>{translated.customTheme}</h2>
        <Demo2 />
        <h2>{translated.title1}</h2>
        <Demo3 />
        <h2>{translated.title2}</h2>
        <Demo4 />
        <h2>RTL</h2>
        <Demo5 />
      </div>
    </>
  )
}

export default ConfigProviderDemo
