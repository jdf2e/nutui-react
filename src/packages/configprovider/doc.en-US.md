# ConfigProvider

### Introduce

Used to configure NutUI-React components globally, providing internationalization support.

### Install

``` javascript
import { ConfigProvider } from '@nutui/nutui-react';
```

## Demo

:::demo

```tsx
import React from 'react';
import { ConfigProvider, Textarea } from "@nutui/nutui-react";
import en from "@nutui/nutui-react/dist/locales/en-US";

const App = () => {
  return (
    <ConfigProvider locale={en}>
      <Textarea />
    </ConfigProvider>
  )
}

export default App;
```

:::

## API

### Props

| Props    | Description                             | Type   | Default          |
|--------------|----------------------------|--------|-----------------|
| locale         | Set up multilingual packs                     | BaseLang | zhCN                |
