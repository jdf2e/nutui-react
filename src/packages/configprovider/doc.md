#  ConfigProvider组件

### 介绍

基于 xxxxxxx

### 安装



## 代码演示

### 基础用法1



## API

### Props

| 参数         | 说明                             | 类型   | 默认值           |
|--------------|----------------------------------|--------|------------------|
| name         | 图标名称或图片链接               | String | -                |
| color        | 图标颜色                         | String | -                |
| size         | 图标大小，如 '20px' '2em' '2rem' | String | -                |
| class-prefix | 类名前缀，用于使用自定义图标     | String | 'nutui-iconfont' |
| tag          | HTML 标签                        | String | 'i'              |

### Events

| 事件名 | 说明           | 回调参数     |
|--------|----------------|--------------|
| click  | 点击图标时触发 | event: Event |
