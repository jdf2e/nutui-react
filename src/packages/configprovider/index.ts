import {
  ConfigProvider,
  useConfig,
  setDefaultConfig,
  getDefaultConfig,
  useRtl,
} from './configprovider'

export type {
  ConfigProviderProps,
  ConfigProviderDirection,
} from './configprovider'
export type { NutCSSVariables } from './types'
export { useConfig, setDefaultConfig, getDefaultConfig, useRtl }
export default ConfigProvider
