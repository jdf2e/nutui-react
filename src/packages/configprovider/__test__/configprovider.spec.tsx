import * as React from 'react'

import { render } from '@testing-library/react'
import '@testing-library/jest-dom'
import zhTW from '@/locales/zh-TW'
import enUS from '@/locales/en-US'

import { ConfigProvider, useConfig, setDefaultConfig } from '../configprovider'

describe('configprovider', () => {
  let container: any

  beforeEach(() => {
    container = document.createElement('div')
    document.body.appendChild(container)
  })

  afterEach(() => {
    document.body.removeChild(container)
    container = null
  })

  test('should match snapshot', () => {
    const { container } = render(
      <ConfigProvider className="aa" style={{ margin: 8 }}>
        测试
      </ConfigProvider>
    )
    expect(container.firstChild?.nodeName).toBe('DIV')
    expect(container).toMatchSnapshot()
  })

  test('should setDefault correctly', () => {
    setDefaultConfig({
      locale: zhTW,
      theme: {
        nutuiColorPrimary: 'red',
      },
    })
    const Children: React.FC = () => {
      const { locale } = useConfig()
      return <>{locale.confirm}</>
    }
    const { container } = render(
      <ConfigProvider>
        <Children />
      </ConfigProvider>
    )

    const ele = container.querySelector('.nut-configprovider')
    expect(ele).toHaveTextContent('確認')
    expect(ele).toHaveClass('nut-configprovider')
    expect(ele).toHaveStyle('--nutui-color-primary: red')
  })

  test('should theme variable and locale variable injection correctly', () => {
    const Children: React.FC = () => {
      const { locale } = useConfig()
      return <>{locale.save}</>
    }
    const darkTheme = {
      nutuiColorPrimary: 'green',
      nutuiColorPrimaryStop1: 'green',
      nutuiColorPrimaryStop2: 'green',
    }
    const { container } = render(
      <ConfigProvider
        data-testid="configprovider"
        locale={enUS}
        className="bb"
        style={{ margin: 8 }}
        theme={darkTheme}
      >
        <Children />
      </ConfigProvider>
    )

    const ele = container.querySelector('.nut-configprovider')
    expect(ele).toHaveTextContent('Save')
    expect(ele).toHaveClass('nut-configprovider bb')
    expect(ele).toHaveStyle(
      '--nutui-color-primary: green; --nutui-color-primary-stop-1: green; --nutui-color-primary-stop-2: green; margin: 8px;'
    )
  })
})
