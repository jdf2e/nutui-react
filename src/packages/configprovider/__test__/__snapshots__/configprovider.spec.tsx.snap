// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`configprovider > should match snapshot 1`] = `
<div>
  <div
    class="nut-configprovider aa"
    style="margin: 8px;"
  >
    测试
  </div>
</div>
`;
